docs/architecture/adr-001-node-key-refactoring.md

```diff
@@ -0,0 +1,104 @@
+- Remove the `version` field from node body writes.
+- Remove the `leftHash` and `rightHash` fields, and instead store `hash` field.
+- Remove the `orphans` from the tree.
+
+New node structure
+
+```go
+type NodeKey struct {
+    version int64
+    nonce int32
+}
+
+type Node struct {
+	key           []byte
+	value         []byte
+	hash          []byte    // keep it in the storage instead of leftHash and rightHash
+	nodeKey       NodeKey   // new field, the key in the storage
+	leftNodeKey   NodeKey   // new field, need to store in the storage
+	rightNodeKey  NodeKey   // new field, need to store in the storage
+	leftNode      *Node
+	rightNode     *Node
+    size          int64
+	subtreeHeight int8
+	persisted     bool
+}
+```
+
+New tree structure
+
+```go
+type MutableTree struct {
+	*ImmutableTree                                    
+	lastSaved                *ImmutableTree
+	versions                 map[int64]bool           
+	allRootLoaded            bool                     
+	unsavedFastNodeAdditions map[string]*fastnode.Node
+	unsavedFastNodeRemovals  map[string]interface{}   
+	ndb                      *nodeDB
+	skipFastStorageUpgrade   bool 
+
+	mtx sync.Mutex
+}
+```
+
+We will restruct the nonce when save the current version. It will reduce unnecessary checks in CRUD operations of the tree and keep sorted the order of insertion in the LSM tree.
+
+### Migration
+
+We can migrate nodes one by one by iterating the version.
+
+- Iterate the version in order, and get the root node for the specific version.
+- Iterate the tree and assign the `nodeKey` to nodes which the node version equals. 
+
+We will implement the `Import` functionality for the original version.
+
+### Pruning
+
+We introduce a new way to prune old versions.
+
+For example, when a user wants to prune the previous 500 versions every 1000 blocks
+- We iterate the tree based on the `n+501`th root and pick only nodes which the version is in `[(n+1), (n+500)]`.
+- We can delete missing nodes (which don't appear in the above traversing) instantly or re-assign the nonce in order for those nodes. Re-assign should be done after stopping the node but it can lead to improving the data locality.
+
+## Consequences
+
+### Positive
+
+Using the version and the sequenced integer ID, we take advantage of data locality in the LSM tree. Since we commit the sorted data, it can reduce compactions and makes easy to find the key. Also, it can reduce the key and node size in the storage.
+
+Removing orphans also provides performance improvements including memory and storage saving. Also, it makes it easy to rollback the tree. Because we can remove all nodes for which the `nodeKey` is greater than the specified version.
+
+### Negative
+
+It requires extra storage to store the node because it should keep `leftNodeKey` and `rightNodeKey` to iterate the tree. Instead, we can delete`leftHash` and `rightHash` fields in the node and reduce the key size.
```

Comment:
could we add how much extra per key would be added?

Reply:
it will reduce the node size as a result, but it's difficult to say how much because `EncodeVarint` is variant.
add `hash` +32byte
remove `leftHash`, `rightHash` -64byte
remove `version` max -8byte
add `leftNodeKey`, `rightNodeKey` max (8+4) * 2 = 24byte

so we can save 16byte in the node body

remove `hash` -32byte
add `version|nonce` 8 + 4 = 12 byte

so we can save 20byte in the node key